### ISSUE TITLE
How to find a given user (Botium user (test)) in salesforce CRM 

### ISSUE THREAD

**Issue by DiegoMoussallem** [codeforequity-at/botium-connector-einsteinbot#1]
Hi, 

I very much liked this connector, thank you for developing it. I would like to ask you if you have any idea how to find a given user in Salesforce while using botium. For example, our Einsteinbot relies on existing users to start the conversation, in case the user is not found, it basically stops the dialog. So, the botium solution is not working, I have already created a user named BotiumTester in my database, however, I need to pass this parameter via botium to Einstein. I found that I can use the prechatDetails, but it is not working. Would you have any idea how I can do it? 

In my previous projects, I have relied on SIMPLE_REST_BODY_TEMPLATE then passing the customer id, but here it does not work. 

any help is appreciated 

Best,

**Comment by codeforequity-at** [comment]
Currently, the prechatDetails are not filled from this connector. If you tell us how the prechatDetails should look like in your case, I can maybe tell you a way how to configure this connector.

**Comment by DiegoMoussallem** [comment]
Hi @codeforequity-at thank you for your prompt response. I basically check the account email. What I am bit confused is by the prechatEntities, what is this basically? because my javascript looks like this: 

` var initESW = function (gslbBaseURL, produto, assetId) {
            embedded_svc.settings.displayHelpButton = true; //Or false
            embedded_svc.settings.language = 'pt-BR'; //For example, enter 'en' or 'en-US'
            embedded_svc.settings.chatbotAvatarImgURL = "https://devchatbot-grupoglobo.cs196.force.com/resource/EVA_GLOBO_ICONES/Chatbot/AvatarChatbot.png";
            embedded_svc.settings.avatarImgURL = "https://devchatbot-grupoglobo.cs196.force.com/resource/EVA_GLOBO_ICONES/Chatbot/AvatarAtendente.png";
            embedded_svc.settings.smallCompanyLogoImgURL = "https://devchatbot-grupoglobo.cs196.force.com/resource/EVA_GLOBO_ICONES/Chatbot/LogoGloboplay.png";

            embedded_svc.settings.loadingText = 'Carregando...'; //(Defaults to Loading)
            embedded_svc.settings.defaultMinimizedText = 'Iniciar atendimento'; //(Defaults to Chat with an Expert)
            embedded_svc.settings.disabledMinimizedText = 'Atendente offline'; //(Defaults to Agent Offline)
            embedded_svc.settings.enabledFeatures = ['LiveAgent'];
            embedded_svc.settings.entryFeature = 'LiveAgent';

            var chatbutton;
            switch (produto) {
                case "Globoplay":
                    chatbutton = '57303000000Gmhj';
                    break;
                case "GloboplayDisney+":
                    chatbutton = '57303000000KyjV';
                    produto = "Globoplay e Disney+";
                    break;
                case "GloboplayCanaisDisney+":
                    chatbutton = '57303000000Kyja';
                    produto = "Globoplay e Disney+";
                    break;
                case "Disney+":
                    chatbutton = '57303000000KyjV';
                    produto = "Disney+";
                    break;
            }

            embedded_svc.settings.extraPrechatFormDetails = [{
                "label": "Account Email",
                "value": "glbteste.kevyn83@mailinator.com", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Contact Email",
                "value": "glbteste.kevyn83@mailinator.com", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "First Name",
                "name": "FirstName",
                "value": "SEBASTIAO", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Origin",
                "value": "Chat", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Status",
                "value": "Novo", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Produto",
                "value": "Globoplay", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Recordtype",
                "value": "0121N000000r1O3QAI", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "bot",
                "value": "Einstein_Bot",
                "transcriptFields": ["ChatBot__c"],
                "displayToAgent": true
            }, {
                "label": "asset",
                "value": "02i03000000I0tiAAC",
                "transcriptFields": ["AssetID__c"],
                "displayToAgent": true
            }];


            embedded_svc.settings.extraPrechatInfo = [{
                "entityName": "Contact",
                "linkToEntityName": "Case",
                "linkToEntityField": "ContactId",
                "saveToTranscript": "ContactId",
                "entityFieldMaps": [{
                    "isExactMatch": false,
                    "fieldName": "FirstName",
                    "doCreate": false,
                    "doFind": false,
                    "label": "firstName"
                }, {
                    "isExactMatch": false,
                    "fieldName": "LastName",
                    "doCreate": false,
                    "doFind": false,
                    "label": "LastName"
                }, {
                    "isExactMatch": true,
                    "fieldName": "Email",
                    "doCreate": false,
                    "doFind": true,
                    "label": "Contact Email"
                }]
            }, {
                "entityName": "Account",
                "linkToEntityName": "Case",
                "linkToEntityField": "AccountId",
                "saveToTranscript": "AccountId",
                "entityFieldMaps": [{
                    "isExactMatch": true,
                    "fieldName": "EVA_Email__c",
                    "doCreate": false,
                    "doFind": true,
                    "label": "Account Email"
                }]
            }, {
                "entityName": "Case",
                "showOnCreate": true,
                "saveToTranscript": "CaseId",
                "entityFieldMaps": [{
                    "isExactMatch": false,
                    "fieldName": "Subject",
                    "doCreate": true,
                    "doFind": false,
                    "label": "issue"
                }, {
                    "isExactMatch": false,
                    "fieldName": "Status",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Status"
                }, {
                    "isExactMatch": false,
                    "fieldName": "Origin",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Origin"
                }, {
                    "isExactMatch": false,
                    "fieldName": "Type",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Produto"
                }, {
                    "isExactMatch": false,
                    "fieldName": "RecordTypeId",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Recordtype"
                }]
            }];
            embedded_svc.init(
                'https://devchatbot-grupoglobo.cs196.force.com',
                '',
                gslbBaseURL,
                '00D030000008aQ5',
                'Globoplay_Agents',
                {
                    baseLiveAgentContentURL: 'https://c.la2-c1cs-ia4.salesforceliveagent.com/content',
                    deploymentId: '5721N000000LDkA',
                    buttonId: chatbutton,
                    baseLiveAgentURL: 'https://d.la2-c1cs-ia4.salesforceliveagent.com/chat',
                    eswLiveAgentDevName: 'Globoplay_Agents',
                    isOfflineSupportEnabled: false
                }
            );
        };


        var startEinsteinGloboplay = function (produto, assetID) {
            if (!window.embedded_svc) {
                let chat = document.querySelector('#chatbot-area');
                var s = document.createElement('script');
                s.setAttribute('src', 'https://devchatbot-grupoglobo.cs196.force.com/embeddedservice/5.0/esw.min.js');
                s.onload = function () {
                    initESW(null);
                };
                chat.appendChild(s);
            } else {
                initESW('https://service.force.com', produto, assetID);
            }
        }
    `

Basically, I need to pass the following variables: 

`  Initializing context variable: $Context.AccountId = 00103000004ggx6AAA
  Initializing context variable: $Context.ContactId = 00303000003OA7yAAG
  Initializing context variable: $Context.AssetId = 02i03000000I0tiAAC `

The ones below are already implemented as far as I understood. 
` 
 Initializing context variable: $Context.EndUserId = 571030000008WuHAAU
  Initializing context variable: $Context.RoutableId = 570030000008ZBKAA2
  Initializing context variable: $Context.ChatKey = 25773c42-6ce7-45a3-959a-a379a340cf88`

**Comment by DiegoMoussallem** [comment]
Hi Florian, 

I need to pass the following things 

```
  Initializing context variable: $Context.AccountId = 00000
  Initializing context variable: $Context.ContactId = 0000
  Initializing context variable: $Context.AssetId = 0000
```

These are already generated 
```
  Initializing context variable: $Context.EndUserId = 000
  Initializing context variable: $Context.RoutableId = 000
  Initializing context variable: $Context.ChatKey = 000
```

**Comment by codeforequity-at** [comment]
Can you tell me how to set these fields based on the Einstein API description ?

https://developer.salesforce.com/docs/atlas.en-us.live_agent_rest.meta/live_agent_rest/live_agent_rest_request_bodies.htm

**Comment by DiegoMoussallem** [comment]
Hey, I use the following JSON in my javascript:

```
            PrechatDetails = [{
                "label": "Account Email",
                "value": "teste@gmail.com", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Contact Email",
                "value": "teste@gmail.com", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "First Name",
                "name": "FirstName",
                "value": "Teste", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Origin",
                "value": "Chat", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Status",
                "value": "Novo", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Produto",
                "value": produto, // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "Recordtype",
                "value": "XXXXXX", // set it dynamically from external system if chat hosted in external system
                "displayToAgent": true
            }, {
                "label": "bot",
                "value": "Einstein_Bot",
                "transcriptFields": ["ChatBot__c"],
                "displayToAgent": true
            }, {
                "label": "asset",
                "value": assetId,
                "transcriptFields": ["AssetID__c"],
                "displayToAgent": true
            }];

            prechatEntities = [{
                "entityName": "Contact",
                "linkToEntityName": "Case",
                "linkToEntityField": "ContactId",
                "saveToTranscript": "ContactId",
                "entityFieldMaps": [{
                    "isExactMatch": false,
                    "fieldName": "FirstName",
                    "doCreate": false,
                    "doFind": false,
                    "label": "firstName"
                }, {
                    "isExactMatch": false,
                    "fieldName": "LastName",
                    "doCreate": false,
                    "doFind": false,
                    "label": "LastName"
                }, {
                    "isExactMatch": true,
                    "fieldName": "Email",
                    "doCreate": false,
                    "doFind": true,
                    "label": "Contact Email"
                }]
            }, {
                "entityName": "Account",
                "linkToEntityName": "Case",
                "linkToEntityField": "AccountId",
                "saveToTranscript": "AccountId",
                "entityFieldMaps": [{
                    "isExactMatch": true,
                    "fieldName": "EVA_Email__c",
                    "doCreate": false,
                    "doFind": true,
                    "label": "Account Email"
                }]
            }, {
                "entityName": "Case",
                "showOnCreate": true,
                "saveToTranscript": "CaseId",
                "entityFieldMaps": [{
                    "isExactMatch": false,
                    "fieldName": "Subject",
                    "doCreate": true,
                    "doFind": false,
                    "label": "issue"
                }, {
                    "isExactMatch": false,
                    "fieldName": "Status",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Status"
                }, {
                    "isExactMatch": false,
                    "fieldName": "Origin",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Origin"
                }, {
                    "isExactMatch": false,
                    "fieldName": "Type",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Produto"
                }, {
                    "isExactMatch": false,
                    "fieldName": "RecordTypeId",
                    "doCreate": true,
                    "doFind": false,
                    "label": "Recordtype"
                }]
            }];
                }
            );
        };
        }
```

I found this page here -> https://help.salesforce.com/articleView?id=000340657&type=1&mode=1, and I am trying to change the index.js contained in the repo, but it is not working. I also tried to change the botium.json but not working either. I think when I rebuild the einstein-connector, it does not use the local index.js but the npm online version.

**Comment by codeforequity-at** [comment]
I added the required changes to the code. You can now set the prechat details and entities in the botium.json.
This is not yet published on NPMJS. So until we build the next release, you have to install this connector directly from Github instead of NPM:

    npm install https://github.com/codeforequity-at/botium-connector-einsteinbot.git#master



```
{
  "botium": {
    "Capabilities": {
      "PROJECTNAME": "Botium Project Einstein Bot",
      "CONTAINERMODE": "einsteinbot",
      "EINSTEINBOT_ENDPOINT": "xxx",
      "EINSTEINBOT_ORGANIZATION_ID": "xxx",
      "EINSTEINBOT_DEPLOYMENT_ID": "xxx",
      "EINSTEINBOT_CHATBUTTON_ID": "xxx",
      "EINSTEINBOT_PRECHAT_DETAILS": [
        {
          "label": "Account Email",
          "value": "teste@gmail.com",
          "displayToAgent": true
        }, {
          "label": "Contact Email",
          "value": "teste@gmail.com",
          "displayToAgent": true
        }
      ],
      "EINSTEINBOT_PRECHAT_ENTITIES": [{
        "entityName": "Contact",
        "linkToEntityName": "Case",
        "linkToEntityField": "ContactId",
        "saveToTranscript": "ContactId",
        "entityFieldMaps": [{
            "isExactMatch": false,
            "fieldName": "FirstName",
            "doCreate": false,
            "doFind": false,
            "label": "firstName"
        }, {
            "isExactMatch": false,
            "fieldName": "LastName",
            "doCreate": false,
            "doFind": false,
            "label": "LastName"
        }, {
            "isExactMatch": true,
            "fieldName": "Email",
            "doCreate": false,
            "doFind": true,
            "label": "Contact Email"
        }]
      }]
    }
  }
}
```

**Comment by DiegoMoussallem** [comment]
It worked!!! Thank you so much @codeforequity-at

**Comment by codeforequity-at** [comment]
Glad to help!
